### PR TITLE
Modernize course request page

### PR DESCRIPTION
**Pull request by trombonekenny**
Convert instructorRequestCourse to TS
EJS file to html.ts

No new features in this PR, just modernizing

### CODE REVIEW

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts:27]: This is really a nit, but it bothers me that Prettier does not resolve this without a nudge.
```suggestion
      { user_id: res.locals.authn_user.user_id },
```

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:215]: I know this is the text that already existed before, but I wonder if this should be changed to phase out the enroll form, or at the very least point the student to the instructor, since they may be hiding the course and using a link in the LMS/course page.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:50]: You removed the Details column here. It is not in the schema, but it is in the query:
https://github.com/PrairieLearn/PrairieLearn/blob/563c91c35c001618e096b363b66a969aa889dff4/apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.sql#L4-L13

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts:28]: As implied by the other comment, this is the wrong schema for it, as it is missing the `details` column.

**Review by jonatanschroeder — COMMENTED** [review]
I compared locally and other than the missing Details column (which seems to be a bug introduced in a previous version) it all looks no-op to me, so that column is the only hold-off on approval.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:50]: Oh, good catch! I noticed details wasn't in the schema or database so I took it as a feature that never got developed. I didn't think about us manually generating it in the SQL. That seems like logic we should move to the html.ts, right?

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:215]: Yeah, you're right that changing prematurely may not be a good idea, but I'd prefer some adjustments at least to the last sentence. Something like:

> If your course is not listed there, contact your instructor for instructions on how to access your assessments.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:50]: Indeed, you're right, this could be removed from the query and moved to the html.ts file, so a change in the schema may not be necessary.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts:28]: Assuming the logic is moving to the TS file, this schema can remain, so marking as resolved.

Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts:28]: I dunno, maybe it's better to add it to the schema (but then the schema won't exactly match the table it was built from). Or maybe we lose the details like the PR implies.

I'm not sure which is better.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:51]: I made a pass at doing this. We'll see what the rest of the crew things about it.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts
Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts:27]: I've given up trying to fight with the formatters. 😸

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:51]: I didn't realize there was a join with the approver user. In that case, I think it makes sense to either keep the previous query computation, or at least add the approver name as a column to the query, then use that name in the computation. In that case, you can check what is done in:
https://github.com/PrairieLearn/PrairieLearn/blob/563c91c35c001618e096b363b66a969aa889dff4/apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.html.ts#L10-L15

and create a `CourseRequestRowSchema` object that adds the new column.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts:41]: This is an FYI, since I believe the other option (new schema) is superior, but if you need something like this, the new objects in `models` have methods for this kind of search:
https://github.com/PrairieLearn/PrairieLearn/blob/563c91c35c001618e096b363b66a969aa889dff4/apps/prairielearn/src/models/user.ts#L7

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:51]: Yep, let's update the schema to avoid a nasty N+1 query here. The "new" style Matt has been pushing for would be something like this:

```ts
const CourseRequestRowSchema = z.object({
  course_request: CourseRequestSchema,
  approved_by_user: UserSchema,
});
```

Of course, update the query to match. `to_json` is your friend!

Comment (mwest1066) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:51]: Yes! Please switch to this, it's sooooo much cleaner+safer+easier!

Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:51]: It's nice getting rid of the need for COALESCE in the SQL.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts]: Changed in c1a0a26a

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:51]: Implemented in c1a0a26a

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts]: I'd like to copy over the PT convention of putting these "row" schema in the `.html.ts` file. I don't have a strong reason for doing this except that we need to pick a convention and we've already got that one, so let's stick with it.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:16]: Please call this variable `rows`. We reserve the name `course_requests` for arrays of type `CourseRequest` and we use `rows` for the generic name on pages that show rows of things.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:61]: Use destructuring here:
```suggestion
                        ${rows(({ course_request, approved_by_user }) => {
```
This avoids the need to ever refer to the whole row variable.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:34]: For real modernization we'd also remove all use of jQuery 😄. It's fine to leave it for now, but I also wouldn't object to getting rid of it. There are various [cheat sheets](https://tobiasahlin.com/blog/move-from-jquery-to-vanilla-javascript/) on the web for this.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:43]: Put this block into a separate function (at the end of the file) and just include it here. That is, here we'd have something like `${CourseRequestsCard({ rows, csrfToken: resLocals.__csrfToken })}`. Note that we do not pass the entire `resLocals` into these helper functions.

The helper function itself would look like:
```
function CourseRequestsCard({rows, csrfToken}: {rows: CourseRequestRow[], csrfToken: string}): HtmlSafeString {
}
```

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:97]: Also stick this card in a helper function and include it here. This will remove the need for the `<!-- New Request -->` comment.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:12]: Just use a regular function rather than an arrow function.
```suggestion
export function RequestCourse({
```

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts
Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.ts]: Thanks, I wasn't sure where to put it. I moved it to `.html.ts` to avoid circular dependencies.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:61]: I presume you mean `rows.map(({ course_request, approved_by_user }).....`?

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:34]: I'm going to hold off on that because I'm trying to not change the functionality. I feel like jQuery cleanup would want to be more consistent across multiple places where similar patterns are used and I'm trying to keep this PR scoped down.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts]: Changed in 9664c3b5

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts]: Changed in 9664c3b

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (trombonekenny) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts]: Changed in 9664c3b

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:54]: Nit: you can keep things less indented if you drop the unnecessary `else` here.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:53]: Nit: you can return a plain empty string (`''`) and change the return type to `HtmlValue`. This avoids an extra object allocation. I acknowledge that this basically doesn't matter here, but it's a helpful pattern to establish for higher-traffic pages.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:76]: Nit, this shouldn't be necessary.

```suggestion
                      details = `Approved by ${approved_by_user.name}`;
```

**Review by nwalters512 — APPROVED** [review]

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts]: Done in af932a8.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts]: Done in af932a8.

**Pull request merged**
